chips/stm32f303xc/src/adc.rs

```diff
@@ -0,0 +1,740 @@
+        }
+    }
+
+    pub fn set_client<C: EverythingClient>(&self, client: &'static C) {
+        self.client.set(client);
+    }
+    // 12
+    pub fn is_enabled_clock(&self) -> bool {
+        self.clock.is_enabled()
+    }
+
+    pub fn enable_clock(&self) {
+        self.clock.enable();
+    }
+
+    pub fn disable_clock(&self) {
+        self.clock.disable();
+    }
+}
+
+struct AdcClock(rcc::PeripheralClock);
+
+impl ClockInterface for AdcClock {
+    fn is_enabled(&self) -> bool {
+        self.0.is_enabled()
+    }
+
+    fn enable(&self) {
+        self.0.enable();
+    }
+
+    fn disable(&self) {
+        self.0.disable();
+    }
+}
+
+impl hil::adc::Adc for Adc {
+    type Channel = Channel;
+
+    fn sample(&self, channel: &Self::Channel) -> ReturnCode {
+        if self.status.get() == ADCStatus::Idle {
+            self.status.set(ADCStatus::OneSample);
+            self.registers.smpr2.modify(SMPR2::SMP16.val(0b100));
+            self.registers.sqr1.modify(SQR1::L.val(0b0000));
+            self.registers.sqr1.modify(SQR1::SQ1.val(*channel as u32));
+            self.registers.ier.modify(IER::EOSMPIE::SET);
+            self.registers.cr.modify(CR::ADSTART::SET);
+            ReturnCode::SUCCESS
+        } else {
+            ReturnCode::EBUSY
+        }
+    }
+
+    fn sample_continuous(&self, _channel: &Self::Channel, _frequency: u32) -> ReturnCode {
+        ReturnCode::ENOSUPPORT
+    }
+
+    fn stop_sampling(&self) -> ReturnCode {
+        ReturnCode::ENOSUPPORT
+    }
+
+    fn get_resolution_bits(&self) -> usize {
+        12
+    }
+
+    fn get_voltage_reference_mv(&self) -> Option<usize> {
+        Some(3300)
+    }
+}
+
+/// Not yet supported
+impl hil::adc::AdcHighSpeed for Adc {
+    /// Capture buffered samples from the ADC continuously at a given
+    /// frequency, calling the client whenever a buffer fills up. The client is
+    /// then expected to either stop sampling or provide an additional buffer
+    /// to sample into. Note that due to hardware constraints the maximum
+    /// frequency range of the ADC is from 187 kHz to 23 Hz (although its
+    /// precision is limited at higher frequencies due to aliasing).
+    ///
+    /// - `channel`: the ADC channel to sample
+    /// - `frequency`: frequency to sample at
+    /// - `buffer1`: first buffer to fill with samples
+    /// - `length1`: number of samples to collect (up to buffer length)
+    /// - `buffer2`: second buffer to fill once the first is full
+    /// - `length2`: number of samples to collect (up to buffer length)
+    fn sample_highspeed(
+        &self,
+        _channel: &Self::Channel,
+        _frequency: u32,
+        _buffer1: &'static mut [u16],
+        _length1: usize,
+        _buffer2: &'static mut [u16],
+        _length2: usize,
+    ) -> (
+        ReturnCode,
+        Option<&'static mut [u16]>,
+        Option<&'static mut [u16]>,
+    ) {
+        (ReturnCode::ENOSUPPORT, None, None)
+    }
```

Comment:
If high speed is not supported, it seems like it makes more sense to simply not implement the trait. Is there something about how the ADC capsule is written that makes doing so not ergonomic?

Reply:
Couldn't compile ADC and integrate it in the kernel without high speed Hil implementation

Reply:
If we can split the ADC capsule in two different parts, more exactly ADC and High Speed ADC, this would work, otherwise the implementation for High Speed is needed for the capsule to be compiled alongside the kernel.

Reply:
Hmm ok that makes sense. I'm not sure about providing a fake interface to userspace, the ADC capsule still isn't supported on the stm32 at this point. I can imagine being on the other side as a developer and thinking ADC will work and then learning why all of the syscalls are returning errors.

Reply:
I still suggest splitting the ADC capsule in two as this will allow an easier implementation of a virtual ADC

Reply:
I think that the ADC capsule should have been (and still should be) split in two. That didn't happen at the time because of the complexity of implementing virtualization to share one ADC between two interfaces, which remains true today.

I think in lieu of more immediate work on the ADC capsule side, this ENOSUPPORT implementation is a reasonable choice. I would follow this up with a comment in main.rs by the ADC configuration that the highspeed ADC trait is not actually implemented for now, and when you fill out https://github.com/tock/tock/blob/master/chips/README.md#hil-support, I would not mark highspeed ADC as implemented.